### ISSUE TITLE
CTC loss is numerically unstable for long sequence lengths

### ISSUE THREAD

**Issue by andykitchen** [tensorflow/tensorflow#4193]
Hi All, great work TF team, thanks for implementing CTC loss ops recently!

I've noticed that CTC loss becomes quite numerically unstable for long sequences, in the attached code that reproduces the plot from the CTC paper, sequences longer than 10,000 or so degrade the quality of the gradients quite severely.

I'm not sure what the best fix is, trying to improve the numerical stability of CTC loss calculation could help, but I have no idea how difficult this would be.

In the interim, I suppose it might be worth issuing a warning when the op is used with long sequences, or placing internal consistency checks that warn the user when CTC loss may be producing bogus outputs. Or even just improving the documentation for CTC loss so that users know that this is a pitfall.

This is also sort of a public service announcement so nobody gets burned on long sequences.

Again, great work, thanks for all your effort.

:)
### Environment info

Pip Package Version: 0.10.0rc0
### Minimal Example

``` python
import numpy as np
import tensorflow as tf

%matplotlib inline
import matplotlib.pyplot as plt

# n = 100
n = 20000
k = 4

with tf.Session() as sess:
    inputs = tf.zeros((n, 1, k+1), dtype=tf.float32)
    labels_indices = tf.constant([[0, 0], [0, 1], [0, 2], [0, 3], [0, 4]], dtype=tf.int64)
    labels_values  = tf.constant([0, 1, 2, 3, 4], dtype=tf.int32)
    labels = tf.SparseTensor(indices=labels_indices, values=labels_values, shape=(1, k+1))
    sequence_length = np.array([n])

    loss = tf.nn.ctc_loss(inputs, labels, sequence_length)
    g, = tf.gradients(loss, inputs)

    g_v = sess.run(g)

plt.plot(-g_v[:,0,:])
```

**Comment by tatatodd** [comment]
@ebrevdo can you comment on this?  Thanks!

**Comment by ebrevdo** [comment]
@andykitchen can you reference the CTC paper & plot you referred to?

**Comment by andykitchen** [comment]
Hi @ebrevdo the plot I'm talking about is the plot created by the code snipped in the issue.

There's no reference, the instability is pretty obvious once you change n to be on the scale of 10s of thousands. Im not currently aware of any published references analysing the numerical stability of CTC

**Comment by gunan** [comment]
ping!
Any updates here?
Is this still a problem with TensorFlow 1.2?

**Comment by standy66** [comment]
@gunan I can confirm that I see the same problem on TensorFlow 1.4.

**Comment by ebrevdo** [comment]
What happens if you run your lstm with float64 then cast to foat32 for the CTC inputs?  Alternatively set num_inter_op_parallel_threads=1?  Do you still get numerical issues?

**Comment by KingRando95** [comment]
Get Outlook for Android<https://aka.ms/ghei36>

________________________________

**Comment by standy66** [comment]
@ebrevdo sorry it took this long to answer.
I am not running any recurrent layers, just OP's code in `Minimal Example` section. Will try `num_inter_op_parallel_threads=1` tomorrow.

**Comment by standy66** [comment]
@ebrevdo so I've set both `inter_op_parallelism_threads=1, intra_op_parallelism_threads=1` in `tf.ConfigProto`, but I am still having the same problem. Just to be clear: [here](https://github.com/standy66/tf_ctc_loss_unstable/blob/master/tf_ctc_loss_unstable.ipynb) is the notebook I'm running. I believe that plots in `In [5], In [6]` should be the same as `In [3], In [4]`, but they are not.

**Comment by standy66** [comment]
Apparently, cudnnCTCLoss suffers from the same numerical instability in gradients. [Here](https://github.com/standy66/cudnn7-ctc-loss-test/blob/master/example.cpp) is the sample code that I wrote. So far, all three CTC loss implementations that I checked (tf's, cudnn's and warp-ctc) are unsuitable for sequences with length > 5000 because of these bogus gradients.

**Comment by ebrevdo** [comment]
You could try adding float64 support to TF's CTC loss kernel.

**Comment by tensorflowbutler** [comment]
Nagging Assignee: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by MaigoAkisame** [comment]
I just ran into the same problem, with a speech recording longer than 1100 frames reporting "No valid path found". My TF version is 1.4.1.

**Comment by tensorflowbutler** [comment]
Nagging Assignee: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by gunan** [comment]
I will leave this as contributions welcome, if someone would like to contribute, you can try adding float64 support for TF CTC loss kernel as @ebrevdo suggested.

**Comment by aprimostka** [comment]
On request added #31164

**Comment by bhack** [comment]
Can we close this?

**Comment by ebrevdo** [comment]
Yep
